### DISCUSSION TITLE
Sbus trainer port inside External module compartment on Radiomaster TX16S.

### DISCUSSION THREAD

**Discussion by Rivelino911** [EdgeTX/edgetx#1638] [Ideas]
Is possible to activate/make this possible. As external compartment is empty, to be used as in Taranis, to put inside a SBUS Rx.
Not hanging in wires outside from Serial 1 or 2.
I know OTX don't put this option from the beginning, but, maybe you can add it.

**Comment by raphaelcoeffic** [comment]
What would make sense would be to use the S.PORT pin, which has an inverter and a real serial port behind.

**Reply by Rivelino911** [reply]
Thanks.
Ideea is to put an Frsky receiver (paired with student Tx) inside external module bay, and use my Radio  as Master trainer. I saw on youtube videosm this is possible with Taranis.

**Reply by raphaelcoeffic** [reply]
> Ideea is to put an Frsky receiver (paired with student Tx) inside external module bay, and use my Radio as Master trainer. I saw on youtube videosm this is possible with Taranis.

That is understood, indeed.

**Reply by Rivelino911** [reply]
Can you be more explicit? Can that be done? How? I searched all internet for a connection to S.Port from a receiver.
Regards.

**Reply by raphaelcoeffic** [reply]
Well, I’d need to code it, but the hardware would allow it.

**Comment by MikePRoberts** [comment]
This would be great, plus if the system could take a sbus or cppm stream then wireless training could be set up with other radio systems.
There is already the option to use a multiprotocol external module emulating a FrSky Rx.  I could not find how the external multiprotocol module sends the data (i.e. which pin and data format).  If this were the same as Rx sbus or cppm then no hardware mods would be required.  Would just need add a external bay option of SBUS RX (of CPPM RX) and change the trainer option of Master/Multi to Master/External (that is in any case clearer).
Is this at a point it could be added as an issue/enhancement?

**Reply by pfeerick** [reply]
I think this PR starts the ball rolling in that direction... it is for the serial ports, but could be the start of this happening in the module bay, hardware capabilities permitting. https://github.com/EdgeTX/edgetx/pull/1462

**Comment by raphaelcoeffic** [comment]
> There is already the option to use a multiprotocol external module emulating a FrSky Rx. I could not find how the external multiprotocol module sends the data (i.e. which pin and data format).

There are basically 2 options for the MPM trainer link:
- via special telemetry packets generate the MPM
- via CPPM on the heartbeat pin (requires a hardware mod on the MPM side)

**Reply by MikePRoberts** [reply]
Many thanks - so CPPM is an option.
However it looks like S.Bus is more readily available from receivers and hence your suggestion to use the S-Port pin looks very attractive.
I hope this can become a reality.
For the time being I have put a 2N7000 N-MOSFET in the lead to AUX1 and it is working.  The 40k ohms internal pullup is a bit weak delivering rise time around 1-2us but appears OK.

**Reply by raphaelcoeffic** [reply]
After looking at my old code back to OpenTx 2.2, it seems the same possibilities can be offered on the same HEARTBEAT pin. However, S.BUS will be on soft-serial on that pin.

**Reply by raphaelcoeffic** [reply]
Faster serial protocols will have to be run on other pins, as it will not work with anything faster than 115.2kbps.

**Reply by MikePRoberts** [reply]
The S-Port pin is surely an easy win?  It has the inverter and code written for Aux1 UART would presumably be applicable to the S-Port UART.

**Reply by MaxStara** [reply]
> After looking at my old code back to OpenTx 2.2, it seems the same possibilities can be offered on the same HEARTBEAT pin. However, S.BUS will be on soft-serial on that pin.

Hi Raphael, any news on this possibility? The feature (already possible in the taranis and opentx) is a great opportunity to have a wireless trainer with any radio that has a ppm output.
![trainer](https://user-images.githubusercontent.com/123209888/213807450-a2e2db0b-ae63-48c9-b8cb-cec29f742924.jpg)

**Reply by raphaelcoeffic** [reply]
@MaxStara which radio is that left on the picture? X9D+ (2019?) or X9-lite?

**Reply by MaxStara** [reply]
hi, is the first Taranis x9d
![IMG_20230128_091105](https://user-images.githubusercontent.com/123209888/215255817-3430af5d-db6e-42d1-ab71-fda2127c1657.jpg)

**Comment by TranquilleXav** [comment]
Hello,
is it possible to implement this solution to do wireless training (as it exists in opentx)?
https://www.tranquille-modelisme.fr/opentx-ecolage-sans-fil-en-sbus-entre-radiomaster-tx16s-et-futaba.html
thank you very much (excuse my poor english! i'm old french...)

**Comment by pierrotm777** [comment]
Hello,

What news about the SBUS input possibility for use as trainer input please ?
Thanks,

Pierre

**Comment by mha1** [comment]
Please check out this PR: https://github.com/EdgeTX/edgetx/pull/4668 
It allows to select Master/SBUS Module (SBUS in S.Port pin)

Find firmware for your radio here: https://github.com/EdgeTX/edgetx/actions/runs/8010206735/artifacts/1267734380

**Reply by TranquilleXav** [reply]
Thank you i test this as soon as possible it's a good news 🙂

**Reply by mha1** [reply]
The PR was merged into 2.10. Use the latest Nightly for testing: https://github.com/EdgeTX/edgetx/releases

**Comment by TranquilleXav** [comment]
thank you, it's ok with TX16s, firmware 2.10.0-rc3, multimodule 1.3.4.0.
a little remarque:

when I turn off the trainer radio, the master radio does not report the loss of trainer
when I turn the trainer radio back on, the master radio does not report the return of trainer.
when i trun off the trainer radio, and turn off the master radio, master radio message:
"trainer still connected"

if you have time for this , thank you very much

**Reply by raphaelcoeffic** [reply]
Are you sure the receiver is set to stop sending anything when it looses signal? It is probably some "failsafe" settings.

**Reply by TranquilleXav** [reply]
probably.
I'm testing this soon...
thank you

**Comment by TranquilleXav** [comment]
I correct my remark:
for receiver SF700S all is ok
it's with receiver R820SF  that it's not ok ...
